x/evm/keeper/grpc_query.go

```diff
@@ -234,6 +236,36 @@ func (q Querier) ConfirmationHeight(c context.Context, req *types.ConfirmationHe
 	return &types.ConfirmationHeightResponse{Height: height}, nil
 }
 
+// GatewayTxState implements the gateway tx state grpc query
+func (q Querier) GatewayTxState(c context.Context, req *types.GatewayTxStateRequest) (*types.GatewayTxStateResponse, error) {
+
+	ctx := sdk.UnwrapSDKContext(c)
+	if !q.keeper.HasChain(ctx, req.Chain) {
+		return nil, status.Error(codes.NotFound, sdkerrors.Wrap(types.ErrEVM, fmt.Sprintf("%s is not a registered chain", req.Chain)).Error())
+	}
+
+	var pollState vote.PollState
+
+	pollKey := vote.NewPollKey(types.ModuleName, fmt.Sprintf("%s_%s", req.Chain, req.TxID.Hex()))
+	poll := q.voter.GetPoll(ctx, pollKey)
+	switch {
+	case poll.Is(vote.NonExistent):
+		pollState = vote.NonExistent
+	case poll.Is(vote.Completed):
+		pollState = vote.Completed
+	case poll.Is(vote.Expired):
+		pollState = vote.Expired
+	case poll.Is(vote.Failed):
+		pollState = vote.Failed
+	case poll.Is(vote.Pending):
+		pollState = vote.Pending
+	default:
+		return nil, status.Error(codes.Internal, sdkerrors.Wrap(types.ErrEVM, "unknown state type").Error())
+	}
```

Comment:
this shouldn't query the poll state but the state of the tx (which should be provided by the keeper without this brittle switch)

Reply:
general message parsing is to be completed soon. This will likely enable us to implement the query differently. For now, marking the query as a candidate for future improvements, and exposing `Poll`'s `PollState` for a more convenient implementation.